Aikido.Zen.Core/Models/AgentContext.cs

```diff
@@ -69,56 +79,63 @@ public void AddHostname(string hostname)
                 host.Port = portNumber;
 
             var key = $"{name}:{port}";
-            _hostnames[key] = host;
+            // thread safe add or update
+            _hostnames.AddOrUpdate(key, host, (_, __) => host);
         }
 
         public void AddUser(User user, string ipAddress)
         {
             if (user == null)
                 return;
-            if (!_users.TryGetValue(user.Id, out UserExtended userExtended))
-            {
-                userExtended = new UserExtended(user.Id, user.Name)
+            _users.AddOrUpdate(
+                user.Id,
+                id => new UserExtended(id, user.Name)
                 {
                     FirstSeenAt = DateTimeHelper.UTCNowUnixMilliseconds(),
-                };
-                _users.Add(user.Id, userExtended);
-            }
-            userExtended.LastIpAddress = ipAddress;
-            userExtended.LastSeenAt = DateTimeHelper.UTCNowUnixMilliseconds();
+                    LastIpAddress = ipAddress,
+                    LastSeenAt = DateTimeHelper.UTCNowUnixMilliseconds()
+                },
+                (_, existing) =>
+                {
+                    existing.LastIpAddress = ipAddress;
+                    existing.LastSeenAt = DateTimeHelper.UTCNowUnixMilliseconds();
+                    return existing;
+                }
+            );
         }
 
         public void AddRoute(Context context)
         {
             if (context == null || context.Url == null) return;
-            _routes.TryGetValue(context.Url, out Route route);
-            if (route == null)
-            {
-                route = new Route
+            // thread safe add or update
+            _routes.AddOrUpdate(
+                context.Url,
+                url => new Route
                 {
-                    Path = context.Url,
+                    Path = url,
                     Method = context.Method,
-                    ApiSpec = OpenAPIHelper.GetApiInfo(context)
-                };
-                _routes.Add(route.Path, route);
-            }
-            else
-            {
-                OpenAPIHelper.UpdateApiInfo(context, route, EnvironmentHelper.MaxApiDiscoverySamples);
-            }
-            route.Hits++;
+                    ApiSpec = OpenAPIHelper.GetApiInfo(context),
+                    Hits = 1
+                },
+                (_, existing) =>
```

Comment:
would make it a little bit clearer that this is a route that already existed where we will now merge the api specs